libbpf-tools/bashreadline.bpf.c

```diff
@@ -0,0 +1,38 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/* Copyright (c) 2021 Kui-Feng Lee */
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include "bashreadline.h"
+
+#define TASK_COMM_LEN 16
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(__u32));
+	__uint(value_size, sizeof(__u32));
+} events SEC(".maps");
+
+SEC("uretprobe/readline")
+int BPF_KRETPROBE(printret) {
+	struct str_t data;
+	char comm[TASK_COMM_LEN];
+	u32 pid;
+
+	if (!PT_REGS_RC(ctx))
+		return 0;
+
+	bpf_get_current_comm(&comm, sizeof(comm));
+	if (comm[0] != 'b' || comm[1] != 'a' || comm[2] != 's' || comm[3] != 'h' || comm[4] != 0 )
+		return 0;
+
+	pid = bpf_get_current_pid_tgid() >> 32;
+	data.pid = pid;
+	bpf_probe_read_user(&data.str, sizeof(data.str), (void *)PT_REGS_RC(ctx));
```

Comment:
see my other comment, if it has to be `bpf_probe_read_user()`, then zero termination was correct, sorry about that. But I think it should be `bpf_probe_read_user_str()`. Can you please check if the string is a proper C string, or it's actually a collection of zero-separated strings?

Reply:
bpf_probe_read_user_str() works well.  When a user types 80 or more chars, it cuts the string to 79 bytes and terminates it with a null char.   With bpf_probe_read_user() will return a string of 80 bytes without a null char.
It just happens to work while the memory is clean.

And, yes, it is a proper C string.